lib/policies/msRestNodeJsUserAgentPolicy.ts

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+import * as os from "os";
+import { HttpRequest } from "../httpRequest";
+import { HttpResponse } from "../httpResponse";
+import { BaseRequestPolicy, RequestPolicy } from "../requestPolicy";
+import { RequestPolicyFactory } from "../requestPolicyFactory";
+import { RequestPolicyOptions } from "../requestPolicyOptions";
+import { Constants } from "../util/constants";
+
+/**
+ * Get a RequestPolicyFactory that creates adds the ms-rest user agent to outgoing requests.
+ * @param userAgentInfo The string[] of userAgent details to apply to each outgoing request.
+ */
+export function msRestNodeJsUserAgentPolicy(userAgentInfo: string[]): RequestPolicyFactory {
+    return (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+        return new MsRestNodeJsUserAgentPolicy(userAgentInfo, nextPolicy, options);
+    };
+}
+
+class MsRestNodeJsUserAgentPolicy extends BaseRequestPolicy {
+    constructor(private readonly _userAgentInfo: string[], nextPolicy: RequestPolicy, options: RequestPolicyOptions) {
+        super(nextPolicy, options);
+    }
+
+    send(request: HttpRequest): Promise<HttpResponse> {
```

Comment:
Are we making sure that this filter will only be added or executed for node.js and not the browser in some other way?

Reply:
I'm tempted to go "laissez faire" here--if a browser user tries to do it, we do whatever we do and accept that it won't work in the browser. Otherwise you add a check for the "isNode" flag, flag, someone shows up with a new javascript runtime environment (e.g. react native?) where you can set the user-agent and now we're failing unnecessarily.

Reply:
If a user wants to add this policy to their pipeline, then that's their choice. When we are constructing the default pipeline (hasn't been coded up yet), we'll then check once whether isNode() or not. If we're running in Node, then we'll add the Node-specific functionality, such as adding this policy. If not, then we'll leave it out.

Reply:
Do we plan to extract common functionality into a base userAgentPolicy and then have node and browser specific policies for user agent?
The name of the policy (msRestNodeJsUserAgentPolicy.ts) indicates that this is intended to be used for node.js. Then why not add the safety check `isNode()` in the definition itself?

Reply:
In this way the customer will not make errors as well.

Reply:
It would probably be a good idea to extract the common functionality into the existing userAgentPolicy type. I didn't do it in this PR, but I can in a future one.
I didn't want to add the isNode() safety check because I don't want it to be executed for every request. I'd rather have some sort of error occur because a customer configured their pipeline incorrectly, rather than silently do extra work.